cmd/zc_traverser_blob.go

```diff
@@ -65,24 +68,34 @@ func (t *blobTraverser) isDirectory(isSource bool) bool {
 	return !isSingleBlob
 }
 
-func (t *blobTraverser) getPropertiesIfSingleBlob() (*azblob.BlobGetPropertiesResponse, bool, error) {
+func (t *blobTraverser) getPropertiesIfSingleBlob() (props *azblob.BlobGetPropertiesResponse, isBlob bool, isDirStub bool, err error) {
 	blobURL := azblob.NewBlobURL(*t.rawURL, t.p)
-	blobProps, blobPropertiesErr := blobURL.GetProperties(t.ctx, azblob.BlobAccessConditions{})
+	props, err = blobURL.GetProperties(t.ctx, azblob.BlobAccessConditions{})
 
 	// if there was no problem getting the properties, it means that we are looking at a single blob
-	if blobPropertiesErr == nil && !gCopyUtil.doesBlobRepresentAFolder(blobProps.NewMetadata()) {
-		return blobProps, true, blobPropertiesErr
+	if err == nil {
+		if gCopyUtil.doesBlobRepresentAFolder(props.NewMetadata()) {
+			return props, false, true, nil
+		}
+
+		return props, true, false, err
 	}
 
-	return nil, false, blobPropertiesErr
+	return nil, false, false, err
 }
 
 func (t *blobTraverser) traverse(preprocessor objectMorpher, processor objectProcessor, filters []objectFilter) (err error) {
 	blobUrlParts := azblob.NewBlobURLParts(*t.rawURL)
 	util := copyHandlerUtil{}
 
+	// trim away the trailing slash before we check whether it's a single blob
+	// so that we can detect the directory stub in case there is one
+	hadTrailingSlash := strings.HasSuffix(blobUrlParts.BlobName, common.AZCOPY_PATH_SEPARATOR_STRING)
+	blobUrlParts.BlobName = strings.TrimSuffix(blobUrlParts.BlobName, common.AZCOPY_PATH_SEPARATOR_STRING)
+	*t.rawURL = blobUrlParts.URL()
```

Comment:
It feels odd to trim this _here_.  In particular, the other usage of t.rawURL is isDirectory.  I can't think through in my head whether isDirectory is called before or after this trimming.  I did note that that isDirectory in turn calls getPropertiesIfSingleBlob AND isDirectory relies on the trailing slashes.

Might be be easier to reason about if you moved the stripping to getPropertiesIfSingleBlob.... and did not ever mutate t.rawURL (i.e. leave the / on it).

I'm not absolutely sure there's a problem here - just that it's weird to see one part of the code trimming a suffix when another depends on it being there. 

*** Also *** can you take a look at that dependency on trailing slashes in urlIsContainerOrVirtualDirectory, and see how that releates to the above ****

Reply:
It doesn't strictly seem like t.rawURL is re-used, unless t.traverse is called somehow a second time. Though, I agree that this modifies the state of the traverser, and should be localized to the get properties call.

---
In reply to: [433614556](https://github.com/Azure/azure-storage-azcopy/pull/1008#discussion_r433614556) [](ancestors = 433614556)

Reply:
Ok!